crates/sui-adapter/src/programmable_transactions/execution.rs

```diff
@@ -93,6 +93,78 @@ fn execute_command<E: fmt::Debug, S: StorageView<E>>(
     command: Command,
 ) -> Result<(), ExecutionError> {
     let results = match command {
+        Command::MakeMoveVec(tag_opt, args) if args.is_empty() => {
+            let Some(tag) = tag_opt else {
+                invariant_violation!(
+                    "input checker ensures if args are empty, there is a type specified"
+                );
+            };
+            let ty = context
+                .session
+                .load_type(&tag)
+                .map_err(|e| context.convert_vm_error(e))?;
+            let abilities = context
+                .session
+                .get_type_abilities(&ty)
+                .map_err(|e| context.convert_vm_error(e))?;
+            let type_ = RawValueType::Loaded {
+                ty,
+                abilities,
+                used_in_non_entry_move_call: false,
+            };
+            // BCS layout for any empty vector should be the same
+            let bytes = bcs::to_bytes::<Vec<u8>>(&vec![]).unwrap();
+            vec![Value::Raw(type_, bytes)]
+        }
+        Command::MakeMoveVec(tag_opt, args) => {
+            let mut res = vec![];
+            leb128::write::unsigned(&mut res, args.len() as u64).unwrap();
+            let mut arg_iter = args.into_iter().enumerate();
+            let (mut used_in_non_entry_move_call, tag) = match tag_opt {
+                Some(tag) => (false, tag),
+                // If no tag specified, it _must_ be an object
+                None => {
+                    // empty args covered above
+                    let (idx, arg) = arg_iter.next().unwrap();
+                    let obj: ObjectValue = context.take_arg(CommandKind::MakeMoveVec, idx, arg)?;
+                    res.extend(obj.to_bcs_bytes());
+                    let tag = TypeTag::Struct(Box::new(obj.type_.clone()));
+                    (obj.used_in_non_entry_move_call, tag)
+                }
+            };
+            let elem_ty = context
+                .session
+                .load_type(&tag)
+                .map_err(|e| context.convert_vm_error(e))?;
+            let elem_abilities = context
+                .session
+                .get_type_abilities(&elem_ty)
+                .map_err(|e| context.convert_vm_error(e))?;
+            for (idx, arg) in arg_iter {
+                let value: Value = if elem_abilities.has_copy() {
+                    context.clone_arg(idx, arg)?
+                } else {
+                    context.take_arg(CommandKind::MakeMoveVec, idx, arg)?
+                };
```

Comment:
It's this kind of conditional that I was trying to avoid with one of the suggestions I made on an earlier PR: I think there should only be `context.take_arg`, and that should check whether the arg at the given index has copy, and behave like `clone` if it does, and take if not.

Reply:
Will fix in the next PR